File: mobile/src/main/java/treehou/se/habit/ui/sitemaps/sitemaplist/SitemapListFragment.java
package treehou.se.habit.ui.sitemaps.sitemaplist;

import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v4.util.Pair;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import se.treehou.ng.ohcommunicator.connector.models.OHServer;
import se.treehou.ng.ohcommunicator.connector.models.OHSitemap;
import treehou.se.habit.R;
import treehou.se.habit.module.HasActivitySubcomponentBuilders;
import treehou.se.habit.mvp.BaseDaggerFragment;
import treehou.se.habit.ui.adapter.SitemapListAdapter;
import treehou.se.habit.ui.sitemaps.sitemap.SitemapFragment;

public class SitemapListFragment extends BaseDaggerFragment<SitemapListContract.Presenter> implements SitemapListContract.View {

    private static final String TAG = "SitemapSelectFragment";

    public static final String ARG_SHOW_SITEMAP = "showSitemap";

    @Inject SitemapListContract.Presenter presenter;

    @BindView(R.id.list) RecyclerView listView;
    @BindView(R.id.empty) TextView emptyView;

    private SitemapListAdapter sitemapAdapter;
    private Unbinder unbinder;

    /**
     * Create fragment where user can select sitemap.
     *
     * @return Fragment
     */
    public static SitemapListFragment newInstance() {
        SitemapListFragment fragment = new SitemapListFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    /**
     * Load sitemaps for servers.
     * Open provided sitemap if loaded.
     *
     * @param sitemap name of sitemap to load
     * @return Fragment
     */
    public static SitemapListFragment newInstance(String sitemap) {
        SitemapListFragment fragment = new SitemapListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_SHOW_SITEMAP, sitemap);
        fragment.setArguments(args);
        return fragment;
    }

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public SitemapListFragment() {}

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sitemaplist_list, container, false);
        unbinder = ButterKnife.bind(this, view);

        setupActionBar();
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 1);
        listView.setLayoutManager(gridLayoutManager);
        listView.setItemAnimator(new DefaultItemAnimator());

        sitemapAdapter = new SitemapListAdapter();
        sitemapAdapter.setSitemapSelectedListener(new SitemapListAdapter.SitemapSelectedListener() {

            @Override
            public void onSelected(OHServer server, OHSitemap sitemap) {
                presenter.openSitemap(server, sitemap);
            }

            @Override
            public void onErrorSelected(OHServer server) {
                Log.d(TAG, "Reloading server: " + server.getDisplayName());
                presenter.reloadSitemaps(server);
            }
        });
        listView.setAdapter(sitemapAdapter);

        return view;
    }

    /**
     * Open fragment showing sitemap.
     *
     * @param server the server of default sitemap.
     * @param sitemap the name of sitemap to show.
     */
    @Override
    public void showSitemap(OHServer server, OHSitemap sitemap) {
        FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
        fragmentManager.beginTransaction()
                .replace(R.id.page_container, SitemapFragment.newInstance(server, sitemap))
                .addToBackStack(null)
                .commit();
    }

    /**
     * Setup actionbar.
     */
    private void setupActionBar(){
        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if(actionBar != null) actionBar.setTitle(R.string.sitemaps);
    }

    @Override
    public void hideEmptyView(){
        emptyView.setVisibility(View.GONE);
    }

    /**
     * Clears list of sitemaps.
     */
    @Override
    public void clearList() {
        emptyView.setVisibility(View.VISIBLE);
        sitemapAdapter.clear();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void showServerError(OHServer server) {
        sitemapAdapter.setServerState(server, SitemapListAdapter.STATE_ERROR);
    }

    @Override
    public void populateSitemaps(Pair<OHServer, List<OHSitemap>> serverSitemaps){
        OHServer server = serverSitemaps.first;
        List<OHSitemap> sitemaps = serverSitemaps.second;

        for (OHSitemap sitemap : sitemaps) {
            sitemapAdapter.add(server, sitemap);
        }
    }

    @Override
    public SitemapListContract.Presenter getPresenter() {
        return presenter;
    }

    protected void injectMembers(HasActivitySubcomponentBuilders hasActivitySubcomponentBuilders) {
        ((SitemapListComponent.Builder) hasActivitySubcomponentBuilders.getFragmentComponentBuilder(SitemapListFragment.class))
                .fragmentModule(new SitemapListModule(this, getArguments()))
                .build().injectMembers(this);
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/sitemaps/sitemaplist/SitemapListFragment.kt
package treehou.se.habit.ui.sitemaps.sitemaplist

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_sitemaplist_list.*
import se.treehou.ng.ohcommunicator.connector.models.OHServer
import se.treehou.ng.ohcommunicator.connector.models.OHSitemap
import treehou.se.habit.R
import treehou.se.habit.dagger.HasActivitySubcomponentBuilders
import treehou.se.habit.dagger.fragment.SitemapListComponent
import treehou.se.habit.dagger.fragment.SitemapListModule
import treehou.se.habit.mvp.BaseDaggerFragment
import treehou.se.habit.ui.adapter.SitemapListAdapter
import treehou.se.habit.ui.adapter.SitemapListAdapter.ServerState
import treehou.se.habit.ui.sitemaps.sitemap.SitemapFragment
import javax.inject.Inject
import javax.net.ssl.SSLPeerUnverifiedException

/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */
class SitemapListFragment : BaseDaggerFragment<SitemapListContract.Presenter>(), SitemapListContract.View {

    @Inject lateinit var sitemapListPresenter: SitemapListContract.Presenter

    private var sitemapAdapter: SitemapListAdapter? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_sitemaplist_list, container, false)
    }

    private fun setupSitemapListAdapter() {
        val gridLayoutManager = GridLayoutManager(activity, 1)
        listView.layoutManager = gridLayoutManager
        listView.itemAnimator = DefaultItemAnimator()

        sitemapAdapter = SitemapListAdapter()
        sitemapAdapter!!.setSitemapSelectedListener(object : SitemapListAdapter.SitemapSelectedListener {

            override fun onSelected(server: OHServer, sitemap: OHSitemap?) {
                sitemapListPresenter.openSitemap(server, sitemap)
            }

            override fun onErrorSelected(server: OHServer) {
                Log.d(TAG, "Reloading server: " + server.displayName)
                sitemapListPresenter.reloadSitemaps(server)
            }

            override fun onCertificateErrorSelected(server: OHServer) {

            }
        })
        listView.adapter = sitemapAdapter
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupActionBar()
        setupSitemapListAdapter()
    }

    /**
     * Open fragment showing sitemap.
     *
     * @param server the server of default sitemap.
     * @param sitemap the name of sitemap to show.
     */
    override fun showSitemap(server: OHServer, sitemap: OHSitemap) {
        val fragmentManager = activity!!.supportFragmentManager
        fragmentManager.beginTransaction()
                .replace(R.id.page_container, SitemapFragment.newInstance(server, sitemap))
                .addToBackStack(null)
                .commit()
    }

    /**
     * Setup actionbar.
     */
    private fun setupActionBar() {
        val actionBar = (activity as AppCompatActivity).supportActionBar
        actionBar?.setTitle(R.string.sitemaps)
    }

    override fun hideEmptyView() {
        emptyView.visibility = View.GONE
    }

    /**
     * Clears list of sitemaps.
     */
    override fun clearList() {
        emptyView.visibility = View.VISIBLE
        sitemapAdapter!!.clear()
    }

    override fun showServerError(server: OHServer, error: Throwable) {
        if (error is SSLPeerUnverifiedException) {
            sitemapAdapter!!.setServerState(server, ServerState.STATE_CERTIFICATE_ERROR)
        } else {
            sitemapAdapter!!.setServerState(server, ServerState.STATE_ERROR)
        }
    }

    override fun populateSitemaps(server: OHServer, sitemaps: List<OHSitemap>) {
        for (sitemap in sitemaps) {
            sitemapAdapter!!.add(server, sitemap)
        }
    }

    override fun getPresenter(): SitemapListContract.Presenter? {
        return sitemapListPresenter
    }

    override fun injectMembers(hasActivitySubcomponentBuilders: HasActivitySubcomponentBuilders) {
        (hasActivitySubcomponentBuilders.getFragmentComponentBuilder(SitemapListFragment::class.java) as SitemapListComponent.Builder)
                .fragmentModule(SitemapListModule(this, arguments!!))
                .build().injectMembers(this)
    }

    companion object {

        private val TAG = "SitemapSelectFragment"

        val ARG_SHOW_SITEMAP = "showSitemap"

        /**
         * Create fragment where user can select sitemap.
         *
         * @return Fragment
         */
        fun newInstance(): SitemapListFragment {
            val fragment = SitemapListFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }

        /**
         * Load sitemaps for servers.
         * Open provided sitemap if loaded.
         *
         * @param sitemap name of sitemap to load
         * @return Fragment
         */
        fun newInstance(sitemap: String): SitemapListFragment {
            val fragment = SitemapListFragment()
            val args = Bundle()
            args.putString(ARG_SHOW_SITEMAP, sitemap)
            fragment.arguments = args
            return fragment
        }
    }
}
